packages/redis/src/RedisStreamSource.ts

```diff
@@ -0,0 +1,142 @@
+import {
+    IInputSource,
+    IRequireInitialization,
+    IDisposable,
+    MessageRef,
+    IComponentContext,
+    DefaultComponentContext,
+    Lifecycle,
+    makeLifecycle,
+} from "@walmartlabs/cookie-cutter-core";
+import { Span, Tags, Tracer } from "opentracing";
+import { generate } from "shortid";
+
+import { IRedisInputStreamOptions, IRedisClient, redisClient } from ".";
+import { RedisOpenTracingTagKeys } from "./RedisClient";
+
+export class RedisStreamSource implements IInputSource, IRequireInitialization, IDisposable {
+    private done: boolean = false;
+    private client: Lifecycle<IRedisClient>;
+    private tracer: Tracer;
+    private spanOperationName: string = "Redis Input Source Client Call";
+    private consumerId: string;
+
+    constructor(private readonly config: IRedisInputStreamOptions) {
+        this.tracer = DefaultComponentContext.tracer;
+        this.consumerId = generate();
```

Comment:
Ah, I think I understand the usage of `XPENDING` a bit better now. I didn't realize we were trying to generate consumer names automatically.

IMO callers should be able to provide their own `consumerId`, especially if they are capable of consistently generating identifiers that will survive restarts (e.g. members of a Kubernetes `StatefulSet`).

Otherwise I think we'll need some way to clean up auto-generated consumers.

Reply:
In talking with Sebastian we felt this was the best approach, at least for an initial implementation, and especially with the above `XPENDING` and `XCLAIM` functionality for recovering and I believe is the same (or very similar) approach the KafkaSource is using.  

What exactly do you mean by 

> Otherwise I think we'll need some way to clean up auto-generated consumers.


>

Reply:
The server will indefinitely retain each auto-generated consumer name as a member of the consumer group.

My primary argument in favor of allowing callers to opt-out of auto-generated consumer names is that it will increase complexity for operational monitoring. Every time a pod is rescheduled or a deployment occurs, we'll have one or more orphaned consumers who's pending messages will need to be reclaimed and then from that point forward the consumer is effectively dead to us. My concern is that it will become non-trivial to determine which of the many consumers in the group *should* be considered active for monitoring purposes.

Reply:
we could def add this to the config (maybe optional and then auto generate it if missing). If the comparison to Kafka holds here, then we actually do the same thing there. The consumerId for Kafka is a GUID we auto generate and the consumer group name is what you configure. Not sure how the cleanup of consumerIds works in Redis ... in Kafka they seem to just disappear after session timeout expired.

Reply:
I would expect the same behavior of redis.  The only state a consumer has, to my understanding, is a string entry on the PEL (Pending Entries List) which will be cleaned up automatically once any consumer starts reading from the consumer group again.  I can definitely make the consumer name configurable, that's not a big deal and an easy change, but I don't think there is anything left to clean up if a consumer dies as they're (essentially) inherently stateless.  Multiple Redis clients can issues `xreadgroup` commands as the same consumer, there is no persistent connection or state to know which client would be the actual consumer.

Reply:
My apologies, I'm likely not doing a great job of explaining my point.

For Kafka consumer groups it is trivial to enumerate all of the partitions being consumed by a group *and* determine if they are actively being consumed from. Furthermore we implicitly expect all partitions to always be consumed. Since Kafka associates consumer checkpoints with partitions, we typically don't even need to know the particular instance/consumer IDs since we can alert on the progress of a partition (because the broker manages consumer partition assignment for us).

However, with Redis Streams we very much so need to care about individual consumers, since we effectively have one logical partition per consumer. If we auto-generate consumer IDs then, unlike Kafka, the set of consumers/logical partitions that we will care about alerting on will *change over time*.

From a monitoring perspective, how will we know if a consumer with 0 pending messages has stalled or has since been replaced? Sure, we can try to infer which one should be considered "active" by comparing the last `seen-time` values, but that's not bulletproof. I would much rather be able to make the assumption that all consumers in the group should be considered "active". I think such an assumption would significantly simplify alerting.

---

As for how consumer cleanup works in Redis, I'm not sure either. The docs don't mention anything about automatic cleanup, only that [`XGROUP DELCONSUMER`](https://redis.io/commands/xgroup) can be used to remove a consumer from a group:

> Consumers in a consumer group are auto-created every time a new consumer name is mentioned by some command. However sometimes it may be useful to remove old consumers since they are no longer used.

However, Antirez's original design [spec](https://gist.github.com/antirez/68e67f3251d10f026861be2d0fe0d2f4#handling-of-consumers) mentions:

> Unlike groups, consumers inside a group are auto-created the first time a given consumer receives a message: when this happens we need to create a pending list new entry referencing the consumer.
>
> However clients may end using different algorithms to create consumer names, and no longer existing consumer names may remain still referenced. However consumers can be reclaimed from memory once the list of pending messages associated with them drops to zero.
>
> The implementation should check periodically the list of consumers in a given group (for instance from time to time when the group is accessed, and only if the number of consumers is already large, or always when the list of consumers is accessed via `XGROUP CONSUMERS`), and if needed reclaim consumers that have both zero associated messages, and a list seen time which is greater than a few minutes.

I haven't read through the source to confirm if or how this was implemented. But my gut feeling is that it would be best not depend on this cleanup logic since the particular thresholds and whatnot are subject to change.